clippy_lints/src/methods/str_split.rs

```diff
@@ -0,0 +1,42 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_context;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+
+use super::STR_SPLIT_AT_NEWLINE;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &'_ Expr<'_>, split_recv: &'_ Expr<'_>, split_arg: &'_ Expr<'_>) {
+    // We're looking for `A.trim().split(B)`, where the adjusted type of `A` is `&str` (e.g. a `str`
+    // literal or an expression returning `String`), and `B` is a `Pattern` that hard-codes a
+    // newline (either `"\n"` or `"\r\n"`). There are a lot of ways to specify a pattern, and
+    // this lint only checks the most basic ones: a `'\n'`, `"\n"`, and `"\r\n"`.
+
+    if let ExprKind::MethodCall(trim_method_name, trim_recv, trim_args, _) = split_recv.kind
+        && trim_method_name.ident.as_str() == "trim"
+        && trim_args.is_empty()
+        && cx.typeck_results().expr_ty_adjusted(trim_recv).peel_refs().is_str()
+        && let ExprKind::Lit(split_lit) = split_arg.kind
+    {
+        let char_n = matches!(split_lit.node, LitKind::Char('\n'));
+        let str_n_or_rn =
+            matches!(split_lit.node, LitKind::Str(sym, _) if (sym.as_str() == "\n" || sym.as_str() == "\r\n"));
+
+        if char_n || str_n_or_rn {
+            let mut app = Applicability::MachineApplicable;
```

Comment:
Due to the behaviour change this should not be `MachineApplicable`

Reply:
Good catch. I had `MaybeIncorrect` in the beginning (for exacty that reason), but then regressed to `MachineApplicable` due to a copy & paste error when I switched from `snippet` to `snippet_with_context`. Fixed.